### ISSUE TITLE
samr21 : RPL_UDP: always falling into "isr_hard_fault"

### ISSUE THREAD

**Issue by maxencechotard** [RIOT-OS/RIOT#2173]
Each time I try to launch RPL_UDP example on samr21 it goes to "isr_hard_fault" even if the program didn't enter in the main where it could execute wrong code. Apparently it goes to "isr_hard_fault" when we call `printf("kernel_init(): This is RIOT! (Version: %s)\n", RIOT_VERSION);` in `kernel_init`.

**Comment by wiredsource** [comment]
@LudwigOrtmann 

We are running with a default stack size of 1.536

**Comment by saurabh984** [comment]
Will this improve things https://github.com/RIOT-OS/RIOT/pull/2155

**Comment by OlegHahm** [comment]
Yepp, merging of #2155 should improve the memory situation for RPL pretty much.

**Comment by OlegHahm** [comment]
**IoT-LAB_M3**
before: 23 808 bytes for stacks
now: 16 640 bytes for stacks

--
**MSB-A2**
before: 25 760 bytes for stacks
now: 17 568 bytes for stacks

--
**native**
before: 204 800 bytes for stacks
now: 147 456 bytes for stacks

--
Please note, that the given numbers do not reflect the actual stack usage.

**Comment by gebart** [comment]
@maxencechotard Can you test if this works now since #2155 has been merged?

**Comment by maxencechotard** [comment]
@gebart I am going to try it

**Comment by maxencechotard** [comment]
@gebart I tried the new version with #2155 and "RPL_MAX_ROUTING_ENTRIES" set to 4 and I get a BSS+data of 31 844. With this version I can initialize a board as node or root without falling into ISR_Hard_Fault. BUT when I try to send a message using "send" command, I fall into ISR_Hard_Fault using "send 0 something". When I try to send a message to a specific node It fail to send my message and I get on the shell "Error Sending Packet !". So I am still working on this problem to get a solution :)

**Comment by maxencechotard** [comment]
@gebart Actually it works when I use send function. I've just forgotten to set RADIO_CHANNEL to a valid channel for at86rf233. So even if my bss is really high  it works and I'll try to reduce my bss with changing my different thread size. There is now just one problem left with the broadcast function of send which induce a ISR_Hard_Fault.

**Comment by maxencechotard** [comment]
I did some optimizations for RIOT threads on samr21 and now I get RPL_UDP example running with BSS+data equal to 30868/32768 with those parameters : 
-RPL_MAX_ROUTING_ENTRIES (12)
-radio thread = 1280 bytes
-rpl_process thread = 1024 bytes
-monitor thread = 1024 bytes

**Comment by gebart** [comment]
This is good to hear. Can you open a PR with the changes necessary to get it working?

**Comment by maxencechotard** [comment]
Yes, I will open a PR this afternoon

**Comment by maxencechotard** [comment]
@gebart Should I modify rpl_config to set RPL_MAX_ROUTING_ENTRIES to 12 when we use the at86rf231 module ? Should I also add in mac.c : 
#ifdef MODULE_AT86RF231
#define RADIO_STACK_SIZE            (1280)
#else
#define RADIO_STACK_SIZE            (KERNEL_CONF_STACKSIZE_MAIN)
#endif

to adjust thread size to samr21 ?

**Comment by gebart** [comment]
I think it is probably better to keep this configuration change linked to the samr21 platform. The Makefile in `sys/net/rpl` supports setting `RPL_MAX_ROUTING_ENTRIES` from the environment.

Try adding something like this to Makefile.include in the samr21 directory:
```make
# samr21 is very memory constrained and does not run with the default 
# RPL_MAX_ROUTING_ENTRIES (= 128)
ifeq (,$(RPL_MAX_ROUTING_ENTRIES))
    RPL_MAX_ROUTING_ENTRIES = 12
endif
```

Regarding the stack size, is it possible to tweak the platform specific parameters further (`KERNEL_CONF_STACKSIZE_DEFAULT` and `KERNEL_CONF_STACKSIZE_PRINTF`) to avoid the `#ifdef` in mac.c?

**Comment by maxencechotard** [comment]
Ok for RPL_MAX_ROUTING_ENTRIES. Regarding the thread optimization It's not possible to optimize mac and rpl threads using (KERNEL_CONF_STACKSIZE_DEFAULT and KERNEL_CONF_STACKSIZE_PRINTF) without changing other thread size which would make other thread size insufficient.

**Comment by gebart** [comment]
Okay,
How about this?
```c
#ifdef RADIO_STACK_SIZE
#define RADIO_STACK_SIZE (KERNEL_CONF_STACKSIZE_MAIN)
#endif
```

**Comment by maxencechotard** [comment]
@gebart Where would add this ifdef ?

**Comment by gebart** [comment]
@maxencechotard I was thinking instead of `#ifdef MODULE_AT86RF231` in mac.c, to avoid hard coding platform specific values in the platform independent source.

**Comment by maxencechotard** [comment]
@gebart Do you mean 

    #ifndef RADIO_STACK_SIZE
    #define RADIO_STACK_SIZE (KERNEL_CONF_STACKSIZE_MAIN)
    #endif

And I'll define RADIO_STACK_SIZE in board.h of samr21-xpro ?

**Comment by gebart** [comment]
Exactly, I accidentally reversed the logic in my example code.

**Comment by maxencechotard** [comment]
PR done

**Comment by maxencechotard** [comment]
Since the BSS is to high with RPL_UDP example, I removed RPL functions and my BSS is now 21744 which is way better to implement an application on SAMR21. But I am now facing a ISR_HARD_FAULT every time I try to debug on the board even if the code is working properly when I manualy reboot the board and follow the execution on the terminal. Is somebody has any idea where this problem could come from ? My branch is :  https://github.com/maxencechotard/RIOT/tree/UDP_ONLY

**Comment by cgundogan** [comment]
@maxencechotard I had a glance at your code, but unfortunately I cannot test this, because I have no hardware available. Unrelated, but the first thing that came to my mind is: how are your devices able to speak with each other? If you remove rpl, then there is nothing that fills up your routing table. I did not see any code in your example related to filling the routing table manually. But I may be missing something here?

You can also remove `char monitor_stack_buffer[MONITOR_STACK_SIZE];` in *rpl.c* since you removed all code that referenced the `monitor_stack` to save some juicy bytes.

**Comment by maxencechotard** [comment]
@cgundogan Actually I filled manually the routing table in ip.c line 114 with setting ipv6_addr_t *dest = &packet->destaddr; and removing condition before it. Well done for monitor stack_buffer :).

**Comment by cgundogan** [comment]
@maxencechotard What I also realized is that you can remove *rpl* from your Makefile `USEMODULE += rpl`. Just to be sure that you do not interact with any part of *rpl* and strip a little bit off the binary size.

**Comment by maxencechotard** [comment]
@cgundogan I removed USEMODULE += rpl and it still goes into ISR_HARD_FAULT.

**Comment by cgundogan** [comment]
Is it working on *native* for you?

**Comment by maxencechotard** [comment]
It works on native

**Comment by maxencechotard** [comment]
The strange thing about that error is that I just fall into ISR_HARD_FAULT when I try to debug. When I launch RIOT without debug on samr21 it works as expected... I would be usually thinking that my RAM is not sufficient but with a BSS of 21k/32k I can't. For me, there is a problem of thread size but I don't understand where and why.

**Comment by gebart** [comment]
Under what circumstances does it trigger hard fault during debug? Is it during program stepping or any other debugger activity?

**Comment by maxencechotard** [comment]
I just reload the code on my board after rebooting this same board and it looks like the bug disappeared :/. I am going to dig this more and give you a feedback if I find something.

**Comment by maxencechotard** [comment]
@gebart It goes to ISR_HARD_FAULT before entering "reset_handler" in startup.c

**Comment by gebart** [comment]
@maxencechotard 
Is the hardware watchdog enabled by default at power on? 

On the Kinetis devices it is not possible to debug until after you disable the hardware watchdog.
I see that there is a line in cpu_init(cpu.c) for disabling the watchdog. Try this in GDB:

```
mon reset halt
thb main
c
```

Does that also end up in hardfault?

One more thing to try:
Move the watchdog disable statement to first thing in the reset handler, it might be running too slow with the debugger enabled to be able to finish all the clib .data copying and .bss clear before hitting cpu_init.

```c
#include "cpu.h"
...
    /* disable the watchdog timer */
    WDT->CTRL.bit.ENABLE = 0;
```

**Comment by maxencechotard** [comment]
@gebart WHat the purpose of those commands ?

    mon reset halt
    thb main
    c

**Comment by gebart** [comment]
reset the device, temporary breakpoint at function `main`, continue execution

**Comment by maxencechotard** [comment]
@gebart I tried your commands and it works after going to ISR_HARD_FAULT but setting "WDT->CTRL.bit.ENABLE" to 1 doesn't works.

**Comment by maxencechotard** [comment]
Any other idea about where the problem come from ?

**Comment by gebart** [comment]
@maxencechotard can you summarize where you are right now by answering these questions?

 - Is flashing the board working properly?
 - Does running the flashed program with debugger disconnected work?
 - Does running the flashed program with debugger connected result in a hard fault?
 - Do you ever hit the breakpoint if you set a breakpoint at `main` in gdb and run?
 - What happens when you run if you set a breakpoint at board_init?

**Comment by maxencechotard** [comment]
Yes for sure.
-The board is flashing correctly even if it is really slow ~4 minutes on a virtual machine with Ubuntu 14.04 LTS on eclipse Luna
-When we run the program on the board without debugging it works perfectly
-Sometimes when we run the same program on the debugger it goes to ISR_HARD_FAULT without being able to enter in "void reset_handler(void)"
-When after entering in ISR_HARD_FAULT and using : 
    mon reset halt
    thb main
    c
The debbuger works and stop in the main. 
-When I set a breakpoint in board_init it also goes to ISR_HARD_FAULT.

Hope those informations will help !

**Comment by gebart** [comment]
Setting a breakpoint at `void reset_handler(void)` will most likely cause the hardware watchdog to reset the CPU, which goes to reset_handler, which causes another debugger break, which resets the CPU, which goes to reset_handler .... etc.

 - How do you normally set your breakpoints? Do you use a GUI or plain command line? Which GDB commands?

Once you get to the breakpoint in `main()` you should be able to single step or set other breakpoints etc. and everything should work as expected. Remember that GDB does not understand the `run` command when using openocd, but rather you should do a `monitor reset halt` (or shortened: `mon reset halt`), followed by a `continue` (or `c`) to do a proper system reset. The `thb` command I listed before sets a temporary breakpoint (one-shot) which will be deleted after you hit it. If you want persistent breakpoints, use the `hb` or `b` commands.

I have no experience with debugging from inside Eclipse (at least I never got it working for anything non-Java) so I won't be able to help there. I have some experience using DDD for complex debugging where a GUI is useful, but I mostly use GDB from the terminal for quick and simple debugging.

**Comment by maxencechotard** [comment]
Usually to set a breakpoint I use breakpoint function in Eclipse witch set a permanent breakpoint at the code line desired. 

Actually, I just try once to set a breakpoint at <code>void reset_handler(void)</code> and every time I fall into <code>ISR_HARD_FALT</code> it does not enter in <code>board_init</code>

**Comment by bapclenet** [comment]
Hi everyone, Kevin Roussel here. I checked Baptiste & Maxence issues on their development machines this morning, and I noticed that the SAMR21 XPro board doesn't seem to be reset after flashing, which usually cause a crash to occur just after a new program is flashed. The board then works normally after a manual reset operation. Since none of us is expert in OpenOCD scripting, could someone check how to reset the SAMR21 board after flashing?

**Comment by thomaseichinger** [comment]
The openocd script actually issues a reset command after flashing. Do you use the before-mentioned patch to speed up flashing? Maybe its worth a try to disable this patch again to be sure. If this doesn't help I could have a closer look at the openocd code.

**Comment by bapclenet** [comment]
@Thomas, thank you for your answer.
We've got the problem with and without the patch so it's not related to the
patch.
 Le 12 févr. 2015 12:52, "Thomas Eichinger" <notifications@github.com> a
écrit :

**Comment by gebart** [comment]
any news on this issue?

Please open a separate issue for the openocd problem with non-reset board after flash.

Is the original crash still happening?

**Comment by bapclenet** [comment]
This crash still happen by using Eclipse to debug. Using debug command line on terminal doesn't make the board crash.

**Comment by OlegHahm** [comment]
This example vanished together with the old network stack.
